### PR TITLE
Add StepwisePlanner extension for MRKL style planning

### PR DESCRIPTION
**Pull request by lemillermicrosoft**
### Motivation and Context
This pull request adds a new extension for semantic planning using a stepwise approach. The StepwisePlanner extension allows users to create and execute plans that consist of a sequence of semantic and native functions, each with a goal and a set of inputs and outputs. The extension uses a semantic search engine to find relevant functions for each step, and a plan creation service to generate a plan that satisfies the user's ask. The extension also provides a system step function that executes the plan and returns the final answer and intermediate observations. The extension can be configured with various parameters, such as the relevancy threshold, the maximum number of relevant functions, the excluded and included functions and skills, and the maximum number of tokens, iterations, and time for the plan.

Regarding #1472

### Description
- Add StepwisePlanner.cs, which registers the planner native functions and the system step function
- Add StepwisePlannerConfig.cs, which defines the configuration options for the StepwisePlanner extension
- Add SystemStep.cs, which represents a step in a Stepwise plan, with properties for the thought, action, action variables, observation, final answer, and original response
- Add helper methods for formatting and validating function views, generating plan requests, and invoking the plan.
- Add logging and error handling for the planner extension
- Add unit tests for the planner extension and the native functions

### Related
- Majority of work initially started from @kaza in #992 

### Changes in other PRs to merge separately
- #1464
- #1465 
- #1466 

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

---------

Co-authored-by: Almir Kazazic <kaza@users.noreply.github.com>

### CODE REVIEW

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+            this._logger?.LogInformation("Ask: {Question}", question);
+            for (int i = 0; i < this.Config.MaxIterations; i++)
+            {
+                var scratchPad = this.CreateScratchPad(question, stepsTaken);
+                this._logger?.LogDebug("Scratchpad: {ScratchPad}", scratchPad);
+                context.Variables.Set("agentScratchPad", scratchPad);
+
+                var llmResponse = (await this._systemStepFunction.InvokeAsync(context).ConfigureAwait(false));
+
+                string actionText = llmResponse.Result.Trim();
```
Comment (lemillermicrosoft) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:111]: Check for error on llmResponse first?

File: dotnet/src/Extensions/Planning.StepwisePlanner/Planning.StepwisePlanner.csproj [hunk 2]
```diff
@@ -0,0 +1,31 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <!-- THIS PROPERTY GROUP MUST COME FIRST -->
+    <AssemblyName>Microsoft.SemanticKernel.Planning.StepwisePlanner</AssemblyName>
+    <RootNamespace>Microsoft.SemanticKernel.Planning.Stepwise</RootNamespace>
```
Comment (shawncal) [inline dotnet/src/Extensions/Planning.StepwisePlanner/Planning.StepwisePlanner.csproj:6]: Note to selves: we need to clean this up for all planners. No reason the assembly, nuget and root namespace shouldn't be identical.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+    {
+        if (stepsTaken.Count == 0)
+        {
+            return string.Empty;
+        }
+
+        var scratchPadLines = new List<string>();
+
+        // Add the original first thought
+        scratchPadLines.Add("This was your previous work (but I haven't seen any of it! I only see what you return as final answer):");
```
Comment (shawncal) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:272]: Where does this string come from? Is it part of some mrkl implementation paper, or just made up? Just wondering, as it seems somewhat informal, and requires a concept of what "you" and "I" are. In the skprompt.txt, "I" seems to be used heavily for the AI persona. Is it the same here ("you" is the caller and "I" is the AI)?

File: dotnet/src/Extensions/Planning.StepwisePlanner/Skills/StepwiseStep/config.json [hunk 4]
```diff
@@ -0,0 +1,32 @@
+{
+  "schema": 1,
+  "description": "Given a request or command or goal generate multi-step plan to reach the goal after each step LLM is called to perform the reasoning for the next step",
```
Comment (shawncal) [inline dotnet/src/Extensions/Planning.StepwisePlanner/Skills/StepwiseStep/config.json:3]: I think this is missing punctuation. Is "After each step..." a new sentence?

File: dotnet/src/Extensions/Planning.StepwisePlanner/Planning.StepwisePlanner.csproj (hunk 2 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/Planning.StepwisePlanner.csproj:6]: https://github.com/orgs/microsoft/projects/866/views/3?pane=issue&itemId=31762117

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 3 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:272]: Made up. Yeah, I changed the wording in the prompt. Will update this sentence to match.

**Pull request merged**